/* Copyright (C) 2014-2015 Alexander Shishenko <alex@shishenko.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
syntax = "proto3";
package librevault.protocol;

message Handshake {
	bytes auth_token = 1;
	string device_name = 2;
	string user_agent = 3;
	repeated string extensions = 4;
}

message HaveMeta {
	bytes path_id = 1;
	int64 revision = 2;
	bytes bitfield = 3;
}

message HaveChunk {
	bytes ct_hash = 1;
}

message MetaRequest {
	bytes path_id = 1;
	int64 revision = 2;
}

message MetaReply {
	bytes meta = 1;
	bytes signature = 2;
	bytes bitfield = 3;
}

message MetaCancel {
	bytes path_id = 1;
	int64 revision = 2;
}

message BlockRequest {
	bytes ct_hash = 1;
	uint32 offset = 2;
	uint32 length = 3;
}

message BlockReply {
	bytes ct_hash = 1;
	uint32 offset = 2;
	bytes content = 3;
}

message BlockCancel {
	bytes ct_hash = 1;
	uint32 offset = 2;
	uint32 length = 3;
}
